PHP 8.4 Support: Property hooks

junichi11 opened the pull request:
- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

### PHP 8.4 Support: Property hooks (Part 1) 

- Fix the lexer and parser(grammar)
- Add `PropertyHookDeclaration` as an `ASTNode`
- Fix `PHP5ErrorHandlerImpl` (handle missing tokens)
- Fix/Add unit tests for the lexer and parser
- Remove `ArrayDimensionSyntaxSuggestionHint` because this no longer works.

Example:
```php
class PropertyHooksClass {
    public int $backed = 100 {
        get {
            return $this->backed;
        }
        set {
            $this->backed = $value;
        }
    }
    public $doubleArrow { // virtual
        get => $this->test();
        set => $this->test() . $value;
    }
    public $attributes {
        #[Attr1] get {}
        #[Attr2] set ($value) {
            $this->attributes = $value;
        }
    }
    public $reference {
        &get => $this->reference;
    }
    public $final {
        final get => $this->final;
    }

    // constructor property promotion
    public function __construct(
        public $prop {
            get {
                return $this->prop;
            }
            set {
                $this->prop = $value;
            }
        }
    ) {}
}

class Child extends PropertyHooksClass {
    public $prop {
        get => parent::$prop::get();
        set {
            parent::$prop::set($value);
        }
    }
}

interface PropertyHooksInterface {
    public string $prop1 {
        get;
    }
    public int $prop2 {
        set;
    }
    public $prop3 {
        get;
        set;
    }
    public $ref { &get; }
}
```

Note: Curly braces array access (`{}` e.g. `$array{1}`, `$array{'key'}`) can use no longer.
Because a conflict occurs in the following case:

```php
"string"{1};
public string $prop = "string" {
    get => $this->prop;
    set {}
}
```
PHP 8.4

![nb-php84-property-hooks-syntax-1](https://github.com/user-attachments/assets/19f742fe-c7ed-4107-91ca-15113e6b40c0)
![nb-php84-property-hooks-syntax-2](https://github.com/user-attachments/assets/8dccb099-e3ee-4563-8381-de5d8d98ebb2)

PHP 8.3

![nb-php84-property-hooks-syntax-php83-1](https://github.com/user-attachments/assets/c08da804-9f98-469b-a03e-005ad3a36276)
![nb-php84-property-hooks-syntax-php83-2](https://github.com/user-attachments/assets/fc566c13-1990-414b-8368-4f43a730bf29)


### PHP 8.4 Support: Property hooks (Part 2) 

- Fix the indexer and the model
- Use JSON format as a signature for property hooks
- Current format is semicolon separated but it's hard to add hooks to a field signature without JSON format
```json
[
    {
        "name":"set",
        "start":3651,
        "end":3690,
        "mod":1,
        "isRef":false,
        "isAttr":false,
        "hasBody":true,
        "paramSig":"$value::0::1:1:0:0:0:0::"
    }
]
```

- Add interface methods
  - `FieldElemnt.isHooked()`
  - `FieldElement.HookedFieldElemnt.isHooked()`
  - `FieldElement.HookedFieldElemnt.getPropertyHooks()`
  - `IndexScope.PHP84IndexScope.findFields()`
  - `TypeScope.FieldDeclarable.getDeclaredFields()`
  - `TypeScope.FieldDeclarable.getInheritedFields()`
  - `PropertyHookScope.isReference()`
  - `PropertyHookScope.hasBody()`
  - `PropertyHookScope.isAttributed()`
  - `PropertyHookScope.getParameterNames()`
  - `PropertyHookScope.getParameters()`
  - `PropertyHookScope.getOffsetRange()`
  - `PropertyHookElement.isReference()`
  - `PropertyHookElement.hasBody()`
  - `PropertyHookElement.isAttributed()`
  - `PropertyHookElement.getParameters()`
  - `PropertyHookElement.getOffsetRange()`

- Add/Fix unit tests for the index and the model


### PHP 8.4 Support: Property hooks (Part 3) 

- Fix the navigator
- Fix/Add unit tests
- Add SVG icons for a property hook, a trait, and an enum case

![nb-php84-property-hooks-navigator](https://github.com/user-attachments/assets/6e15a849-e852-4115-b51a-d262634ab2a5)

![nb-php84-property-hooks-navi-icons](https://github.com/user-attachments/assets/ebf434a4-4f2c-4074-a722-75ee2d143c4b)
![nb-php84-property-hooks-navi-icons-dark](https://github.com/user-attachments/assets/14a8dbe1-0ff9-45ab-812f-c1d517f4cf94)

### PHP 8.4 Support: Property hooks (Part 4) 

- Fix the folding feature
- Add unit tests

Expanded:

![nb-php84-property-hooks-folding-base](https://github.com/user-attachments/assets/a9e2f7c9-a065-4634-8522-b0698c2da928)

Collapsed:

![nb-php84-property-hooks-folding](https://github.com/user-attachments/assets/782572e3-c514-4c2e-bfea-03bbe10f044d)

![nb-php84-property-hooks-folding-options](https://github.com/user-attachments/assets/60cdd93e-ab4f-4f2a-8f48-e52541885212)


### PHP 8.4 Support: Property hooks (Part 5) 

- Fix the formatter and `PhpTypedBreakInterceptor`
- New Formatting Options
  - Braces
    - Field Declaration (Same Line by default)
    - Property Hook (Same Line by default)
  - Blank Lines
    - Between Property Hooks (0 by default)
    - Empty Property Hook Body (Unchecked by default)
  - Spaces
    - Before Left Braces
      - Field Declaration (Checked by default)
      - Property Hook Declaration (Checked by default)
- Add unit tests

Example:
```php
class PropertyHookClass {
    public string $prop1 {get{} set{}}
    public string $prop2 {
        get{return $this->prop2;}
        set{$this->prop2 = $value;}
    }
}

interface PropertyHookInterface {
    public string $prop1 {get; set;}
}
```

Formatted:
```php
class PropertyHookClass {

    public string $prop1 {
        get {}
        set {}
    }
    public string $prop2 {
        get {
            return $this->prop2;
        }
        set {
            $this->prop2 = $value;
        }
    }
}

interface PropertyHookInterface {

    public string $prop1 {
        get;
        set;
    }
}
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 1)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the lexer and parser(grammar)
- Add `PropertyHookDeclaration` as an `ASTNode`
- Fix `PHP5ErrorHandlerImpl` (handle missing tokens)
- Fix/Add unit tests for the lexer and parser
- Remove `ArrayDimensionSyntaxSuggestionHint` because this no longer works.

Example:
```php
class PropertyHooksClass {
    public int $backed = 100 {
        get {
            return $this->backed;
        }
        set {
            $this->backed = $value;
        }
    }
    public $doubleArrow { // virtual
        get => $this->test();
        set => $this->test() . $value;
    }
    public $attributes {
        #[Attr1] get {}
        #[Attr2] set ($value) {
            $this->attributes = $value;
        }
    }
    public $reference {
        &get => $this->reference;
    }
    public $final {
        final get => $this->final;
    }

    // constructor property promotion
    public function __construct(
        public $prop {
            get {
                return $this->prop;
            }
            set {
                $this->prop = $value;
            }
        }
    ) {}
}

class Child extends PropertyHooksClass {
    public $prop {
        get => parent::$prop::get();
        set {
            parent::$prop::set($value);
        }
    }
}

interface PropertyHooksInterface {
    public string $prop1 {
        get;
    }
    final public int $prop2 {
        set;
    }
    public $prop3 {
        get;
        set;
    }
    public $ref { &get; }
}
```

Note: Curly braces array access (`{}` e.g. `$array{1}`, `$array{'key'}`) can use no longer.
Because a conflict occurs in the following case:

```php
"string"{1};
public string $prop = "string" {
    get => $this->prop;
    set {}
}
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 2)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the indexer and the model
- Use JSON format as a signature for property hooks
- Current format is semicolon separated but it's hard to add hooks to a field signature without JSON format
- Use the "JSON simple". If we use "Jackson", problems(timeout, java.lang.NoClassDefFoundError) occurs in CI for Windows

```json
[
    {
        "name":"set",
        "start":3651,
        "end":3690,
        "mod":1,
        "isRef":false,
        "isAttr":false,
        "hasBody":true,
        "paramSig":"$value::0::1:1:0:0:0:0::"
    }
]
```

- Add interface methods
  - `FieldElemnt.isHooked()`
  - `FieldElement.HookedFieldElemnt.isHooked()`
  - `FieldElement.HookedFieldElemnt.getPropertyHooks()`
  - `IndexScope.PHP84IndexScope.findFields()`
  - `TypeScope.FieldDeclarable.getDeclaredFields()`
  - `TypeScope.FieldDeclarable.getInheritedFields()`
  - `PropertyHookScope.isReference()`
  - `PropertyHookScope.hasBody()`
  - `PropertyHookScope.isAttributed()`
  - `PropertyHookScope.getParameterNames()`
  - `PropertyHookScope.getParameters()`
  - `PropertyHookScope.getOffsetRange()`
  - `PropertyHookElement.isReference()`
  - `PropertyHookElement.hasBody()`
  - `PropertyHookElement.isAttributed()`
  - `PropertyHookElement.getParameters()`
  - `PropertyHookElement.getOffsetRange()`

- Add/Fix unit tests for the index and the model

Log:
```
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing finished, indexing took 14 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 1 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 3)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the navigator
- Fix/Add unit tests
- Add SVG icons for a property hook, a trait, and an enum case

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassScopeImpl.java:
```diff
@@ -379,32 +372,145 @@ public Collection<? extends MethodScope> getInheritedMethods() {
         for (ClassScope classScope : superClasses) {
-            Set<org.netbeans.modules.php.editor.api.elements.FieldElement> indexedFields = filterForPrivate.filter(index.getAlllFields(classScope));
-            for (org.netbeans.modules.php.editor.api.elements.FieldElement field : indexedFields) {
-                allFields.add(new FieldElementImpl(classScope, field));
```
This implementation has a bug:
All inherited fields are added to `classScope` as elements. 
If `classScope.getDeclaredFields()` is invoked, inherited fields may also be got.

https://github.com/apache/netbeans/blob/51fa1325cad00b4d9fc728ae207ec9657cabb31e/php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ClassScopeImpl.java#L307-L318

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/verification/ArrayDimensionSyntaxSuggestionHint.java:
This no longer work.

junichi11 commented on php/php.editor/tools/ASTPHP5Parser.cup:
```diff
@@ -5360,23 +5435,13 @@ T_OPEN_RECT:o dim_offset:index T_CLOSE_RECT:end
     RESULT = new ArrayDimension(oleft, endright, index, ArrayDimension.Type.VARIABLE_ARRAY);
 :}
-
-| T_CURLY_OPEN:o dim_offset:index T_CURLY_CLOSE:end
-{:
-    RESULT = new ArrayDimension(oleft, endright, index, ArrayDimension.Type.VARIABLE_HASHTABLE);
-:}
```
Note: Curly braces array access (`{}` e.g. `$array{1}`, `$array{'key'}`) can use no longer.
Because a conflict occurs in the following case:

```php
"string"{1};
public string $prop = "string" {
    get => $this->prop;
    set {}
}
```

tmysik replied:
This should be fine, hopefully.

junichi11 replied:
Yes :)
- deprecated as of PHP 7.4
- fatal error as of PHP 8.0
- parse error as of PHP 8.4

junichi11 commented:
@tmysik This is not done yet. I'm slowly going to add more commits to this PR. Could you please have a look at each commit when you have time?

junichi11 commented:
Time out occurs in Windows tests. I'll try adding commits one by one.

junichi11 commented:
@tmysik Please ignore this. I'll try to investigate this later...

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/model/impl/PropertyHookSignatureItem.java:
```diff
@@ -0,0 +1,298 @@
+            items.addAll(signatureItems);
+            signature = items.toJSONString();
+            if (LOGGER.isLoggable(Level.FINE)) {
+                LOGGER.log(Level.FINE, "getSignatureFromScopes() took: {0} ms", (System.currentTimeMillis() - start)); // NOI18N
```
<details>
<summary>Log</summary>

```
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 1 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing finished, indexing took 19 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing started, time from last indexing 9,124 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 1 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing finished, indexing took 36 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing started, time from last indexing 11,654 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: getSignatureFromScopes() took: 0 ms
INFO [org.netbeans.ui.indexing]: Indexing finished, indexing took 14 ms.
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 1 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
FINE [org.netbeans.modules.php.editor.model.impl.PropertyHookSignatureItem]: fromSignature() took: 0 ms
```
</details>

Example:
```php

<?php
interface InterfaceA {
    public string $a1 {
        get;
    }
    public int $a2 {
        set;
    }
    public $b3 {
        get;
        set;
    }
}

interface InterfaceB extends InterfaceA {
    public string $b1 {
        get;
    }
    public int $b2 {
        set;
    }
    public $b3 {
        get;
        set;
    }
}
interface InterfaceX {
    public string $x1 {
        get;
    }
    public int $x2 {
        set;
    }
    public $x3 {
        get;
        set;
    }
}

interface InterfaceY extends InterfaceB {
    public string $y1 {
        get;
    }
    public int $y2 {
        set;
    }
    public $y3 {
        get;
        set;
    }
}

interface InterfaceZ extends InterfaceX, InterfaceY {
    public string $z1 {
        get;
    }
    final public int $z2 {
        set;
    }
    public $z3 {
        get;
        set;
    }
}

```

tmysik replied:
Looks good :+1:

junichi11 commented:
@tmysik It should be OK now. Probably, the cause is "Jackson". Instead, I used "JSON simple".

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 4)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the folding feature
- Add unit tests

tmysik commented on php/php.editor/src/org/netbeans/modules/php/editor/elements/PropertyHookElementImpl.java:
```diff
@@ -0,0 +1,203 @@
+     * @return property hook elements
+     */
+    public static List<PropertyHookElement> fromSignatureItems(final List<PropertyHookSignatureItem> signatureItems, final String field, final String filenameUrl, final ElementQuery elementQuery) {
+        List<PropertyHookElement> elements = new ArrayList<>();
```
Nitpick:

```suggestion
        List<PropertyHookElement> elements = new ArrayList<>(signatureItems.size());
```

junichi11 replied:
Will fix it.

junichi11 replied:
Fixed: https://github.com/apache/netbeans/compare/33472a7a8c8a28e0896c06e55d656b1e9a9cd648..b7eb3b4cbfc2e915ecc84335650f39ea145074e6

tmysik commented:
I did a _very quick_ review, and it looks good to me. Thanks for changing the code to the modern one, e.g. the switches :)

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/api/elements/FieldElement.java:
```diff
@@ -16,19 +16,56 @@
+                && ((HookedFieldElemnt) field).isHooked();
+    }
+
+    public interface HookedFieldElemnt extends FieldElement {
```
Oops, Thank you for catching it!

junichi11 replied:
Fixed: https://github.com/apache/netbeans/compare/33472a7a8c8a28e0896c06e55d656b1e9a9cd648..b7eb3b4cbfc2e915ecc84335650f39ea145074e6

junichi11 commented:
Fix folding feature: https://github.com/apache/netbeans/pull/8227/commits/3b5198416be956ab0f273638b0a092c542da8758

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 5)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the formatter and `PhpTypedBreakInterceptor`
- New Formatting Options
  - Braces
    - Field Declaration (Same Line by default)
    - Property Hook (Same Line by default)
  - Blank Lines
    - Between Property Hooks (0 by default)
    - Empty Property Hook Body (Unchecked by default)
  - Spaces
    - Before Left Braces
      - Field Declaration (Checked by default)
      - Property Hook Declaration (Checked by default)
- Add unit tests

junichi11 commented on php/php.editor/src/org/netbeans/modules/php/editor/api/PhpModifiers.java:
```diff
@@ -132,8 +132,8 @@ private PhpModifiers(int... bitmask) {
         for (int mode : bitmask) {
             this.mod |= mode;
         }
-        if (!Modifier.isPrivate(mod) && !Modifier.isProtected(mod) && !Modifier.isImplicitPublic(mod)) {
-            mod |= Modifier.PUBLIC;
+        if (!Modifier.isPrivate(mod) && !Modifier.isProtected(mod) && !Modifier.isPublic(mod)) {
+            mod |= Modifier.IMPLICIT_PUBLIC;
         }
```
`IMPLICIT_PUBLIC` is better if `mod` is `0` because the public modifier doesn't exist.

junichi11 commented on php/php.editor/test/unit/data/testfiles/formatting/php84/propertyHooks_01.php.testPropertyHookDeclarationBP_01c.formatted:
```diff
@@ -0,0 +1,544 @@
+            }
+    }
+    public $valid13 {
+        final get
+            {
+            return 100;
+        }
```
This is an existing bug. I didn't fix it (maybe, the priority is very low) because I don't think the `New Line Indented` option is important.

We can also reproduce this using a method.

```php
class C {
    public function example() { return 100;}
}
```

tmysik commented:
@junichi11, please, let me know once this PR is ready for the review (no rush!). Thank you.

junichi11 commented:
@tmysik OK. I'll do that after I fix the other features(Hints, CC, Go to Declaration, etc.) :)

tmysik commented:
@junichi11, as I wrote, take your time, it is a very complex feature. Thank you

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 6)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix the indentation that is inserted when a new line is added
- Fix an existing bug for a lambda function
- Add unit tests

Existing bug:
```php
// example
test(
        function(): void {^
);
```

```php
// before
test(
        function(): void {
    ^
        }
);

// after
test(
        function(): void {
            ^
        }
);
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 7)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix `ModifiersCheckHintError`
- Add unit tests

Example:
```php
class PropertyHooksClass {
    private $invalidPrivateFinal {
        final get{} // cannot use final
    }
    private $invalidPublic01 {
        public get; // cannot use public
    }
    public $invalidStatic01 {
        static get {} // cannot use static
    }
    public static $invalidStatic02 { // cannot use static
        get{}
        set{}
    }
    public $invalidNonAbstract {
        get; // need a body
        set; // need a body
    }
    public readonly int $invalidReadonly { get{} set{} } // cannot use readonly
}

interface PropertyHookInterface {
    abstract public $invalid01 { get; set; } // cannot use abstract
    protected $invalid02 {get; set;} // cannot use protected
    private $invalid03 { // cannot use private
        get;
        set;
    }
    public $invalid04 {
        final get; // cannot use final
    }
    final public $invalid05 { // cannot use final
        get;
        set;
    }
    public $invalid06 {
        get {} // cannot has a body
    }
    public private(set) int $invalid07 { // cannot use private(set)
        set;
    }
    final public int $invalid08 { // cannot use final
        set;
    }
    public readonly int $invalid09 { // cannot use readonly
        get;
    }
}
```

- Fix a formatting issue

Example:
```
interface TestInterface {
    public int $prop1 {get; set;}
    public int $prop2 {}
    ^^^^^^^ selected here
}

// before
interface TestInterface {
    public int $prop1 {get; set;}
public int $prop2 {}
}

// after
interface TestInterface {
    public int $prop1 {get; set;}
    public int $prop2 {}
}
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 8)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix `WrongOrderOfArgsHint`
- Add/Fix unit tests

Example:
```php
    public function __construct(
        private ?int $optional = 1 {
            get {
                return $this->optional;
            }
            set {
                $this->optional = $value;
            }
        },
        $mandatory,
    ) {}

    // apply fix
    public function __construct(
        $mandatory,
        private ?int $optional = 1 {
            get {
                return $this->optional;
            }
            set {
                $this->optional = $value;
            }
        },
    ) {}
```

junichi11 committed:
PHP 8.4 Support: Property hooks (Part 9)

- https://github.com/apache/netbeans/issues/8035
- https://wiki.php.net/rfc#php_84
- https://wiki.php.net/rfc/property-hooks

- Fix `ImmutableVariablesHint`
- Add unit tests

Example:
```php
class PropertyHooksClass {
    public int $backed = 100 {
        set(int $value) {
            $this->backed = $value;
        }
    }
    public $attributes {
        #[Attr2] set ($value) {
            $this->attributes = $value;
        }
    }
}
// before:
class PropertyHooksClass {
    public int $backed = 100 {
        set(int $value) { // a hint is shown
            $this->backed = $value;
        }
    }
    public $attributes {
        #[Attr2] set ($value) { // a hint is shown
            $this->attributes = $value;
        }
    }
}
// after: no hint
```

junichi11 commented:
I'll change the milestone to `NB27`.

junichi11 commented:
Rebased this branch on top of master.
Reviewed: Part 1〜5
Not Reviewed: Part 6〜9

tmysik commented:
@junichi11 great to see you back! :+1:

junichi11 commented:
@tmysik Could you please have a look at this?(Part 6〜9)
This feature isn’t finished yet, but I think we should merge it for now.

tmysik commented:
@junichi11 I quickly reviewed your latest commits (those ones you requested) and I don't see anything suspicious (BTW, thanks for those new constants, with proper types). We do have a lot of tests and all are passing, so let's merge it.

Thank you!

rarila commented:
@junichi11 I Can't wait to try it out! 🎉

It would be nice of you — and perhaps helpful — if you could briefly list what's still missing. Thx!

junichi11 commented:
@tmysik I forgot to push the ready for review, sorry. I'll merge this tomorrow if this is approved. Thank you for the review :)

junichi11 commented:
@rarila e.g. Code completion, Go to declaration, Mark occurrences, ...
You can check what has been implemented by looking at the commit messages☺️ Thanks!

tmysik reviewed (approved):
See my previous messages. @junichi11, thank you for your work!
